### PR TITLE
GH#23551: harden canonical sweep cwd change

### LINKED ISSUE #23551: Review followup: PR #23545 — GH#23543: fix canonical worktree sweep cwd handling
## Unaddressed review bot suggestions

PR #23545 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/23545

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `AGENTS.md`
"AI-Generated Issue Quality"). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `AGENTS.md` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/pulse-canonical-maintenance.sh:345`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

#### gemini-code-assist on `.agents/scripts/pulse-canonical-maintenance.sh:345`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/23545#discussion_r3239373254)

```diff
@@ -340,8 +340,17 @@ _stale_worktree_sweep_single_repo() {
 	# (see _stale_worktree_sweep). That was the ANSI-bleed half of the 2026-04-20
 	# canonical-trash incident. Route output to the logfile instead of /dev/null
 	# so operators can still diagnose sweep failures.
-	if ! _canonical_maintenance_run_with_timeout "$CANONICAL_MAINTENANCE_TIMEOUT" "$worktree_helper" clean --auto --force-merged >>"${LOGFILE:-/dev/null}" 2>&1; then
-		echo "[pulse-canonical-maintenance] Worktree sweep timed out for ${repo_path}" >>"${LOGFILE:-/dev/null}"
+	local sweep_rc=0
+	(
+		cd "$repo_path" || exit 125
```

> medium
> 
> When using `cd` with a variable path, it is a best practice to use the `--` separator to prevent the path from being interpreted as an option if it starts with a hyphen. This improves the robustness of the script against unusual directory names.
> 
> ```suggestion
> cd -- "$repo_path" || exit 125
> ```

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/23545#pullrequestreview-4287625700)

> ## Code Review
> 
> This pull request improves the robustness of the `pulse-canonical-maintenance.sh` script by ensuring the worktree sweep helper runs from the repository's root directory and by using `git rev-parse` to validate git repositories. It also enhances error reporting for sweep failures and adds corresponding unit tests. A suggestion was made to use the `--` separator with `cd` to handle edge-case directory names starting with hyphens.

---
[aidevops.sh](https://aidevops.sh) v3.15.46 automated scan.

### PR DESCRIPTION
**Pull request by marcusquinn**
## Summary

Updated the canonical worktree sweep subshell to use cd -- for variable repo paths, addressing the verified review-followup robustness issue.

## Files Changed

.agents/scripts/pulse-canonical-maintenance.sh

## Runtime Testing

- **Risk level:** Low (agent prompts / infrastructure scripts)
- **Verification:** shellcheck .agents/scripts/pulse-canonical-maintenance.sh; bash .agents/scripts/tests/test-pulse-canonical-maintenance.sh (12 passed)

Resolves #23551

---
[aidevops.sh](https://aidevops.sh) v3.15.46 plugin for [OpenCode](https://opencode.ai) v1.14.50 with gpt-5.5 spent 1m and 37,364 tokens on this as a headless worker.

### CODE REVIEW

**Comment by marcusquinn** [comment]
## Completion Summary

- **What**: Updated the canonical worktree sweep subshell to use cd -- for variable repo paths, addressing the verified review-followup robustness issue.
- **Issue**: #23551
- **Files changed**: .agents/scripts/pulse-canonical-maintenance.sh
- **Testing**: shellcheck .agents/scripts/pulse-canonical-maintenance.sh; bash .agents/scripts/tests/test-pulse-canonical-maintenance.sh (12 passed)
- **Key decisions**: Applied the bot suggestion because the cited line used cd with a variable path and the -- separator is a safe shell robustness improvement.

---
[aidevops.sh](https://aidevops.sh) v3.15.46 plugin for [OpenCode](https://opencode.ai) v1.14.50 with gpt-5.5 spent 2m and 37,364 tokens on this as a headless worker.

**Comment by marcusquinn** [comment]
## Admin Merge Fallback (t2247)

Branch protection blocked the plain `gh pr merge` for PR #23559. The merge succeeded using `--admin` fallback (per GH#18538 — workers share the maintainer's `gh auth`).

**Merge method:** `--squash`

<details>
<summary>Original branch-protection error</summary>

```text
X Pull request marcusquinn/aidevops#23559 is not mergeable: the base branch policy prohibits the merge.
To have the pull request merged after all the requirements have been met, add the `--auto` flag.
To use administrator privileges to immediately merge the pull request, add the `--admin` flag.
```

</details>

**Remediation:** If this bypass was unintended, revert with `gh pr revert 23559 --repo marcusquinn/aidevops` and investigate why review bots did not approve.



---
[aidevops.sh](https://aidevops.sh) v3.15.46 plugin for [OpenCode](https://opencode.ai) v1.14.50 with unknown spent 2m and 38,354 tokens on this as a headless worker.

**Review by alex-solovyev — APPROVED** [review]
Auto-approved by pulse runner @alex-solovyev — author @marcusquinn confirmed collaborator, pre-merge gates passed.

**Comment by alex-solovyev** [comment]
## Completion Summary

- **What**: Updated the canonical worktree sweep subshell to use cd -- for variable repo paths, addressing the verified review-followup robustness issue.
- **Issue**: #23551
- **Files changed**: .agents/scripts/pulse-canonical-maintenance.sh
- **Testing**: shellcheck .agents/scripts/pulse-canonical-maintenance.sh; bash .agents/scripts/tests/test-pulse-canonical-maintenance.sh (12 passed)
- **Key decisions**: Applied the bot suggestion because the cited line used cd with a variable path and the -- separator is a safe shell robustness improvement.

---
[aidevops.sh](https://aidevops.sh) v3.15.46 plugin for [OpenCode](https://opencode.ai) v1.14.50 with gpt-5.5 spent 2m and 37,364 tokens on this as a headless worker.

---
Merged via PR #23559 to main.
_Merged by deterministic merge pass (pulse-wrapper.sh)._
